### PR TITLE
CUDA graph support for TRT EP

### PR DESCRIPTION
**Pull request by chilo-ms**
CUDA EP already supports [CUDA graph](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-graphs), also we observed some models can benefit from using CUDA graph with `trtexec`. Therefore, this PR enables the CUDA graph support for TRT EP.

The implementation is based on https://github.com/microsoft/onnxruntime/pull/9978 with the same [constraints](https://github.com/microsoft/onnxruntime/pull/9978) as below:

- Models with control-flow ops (i.e. If, Loop and Scan ops) are not supported.
- Usage of CUDA Graphs is limited to models where-in all the model ops (graph nodes) can be partitioned to the TRT EP.
- The input/output types of models need to be tensors.
- Shapes of inputs/outputs cannot change across inference calls.
- IObinding is required.

### CODE REVIEW

**Comment by hariharans29** [comment]
Thanks @chilo-ms  @tlwu had a couple of fixes for CUDA Graphs (open PRs) - I was wondering if you needed fixes similar to that while testing your models ?

**Comment by tianleiwu** [comment]
> > Thanks @chilo-ms. @tlwu had a couple of fixes for CUDA Graphs (open PRs) - I was wondering if you needed fixes similar to that while testing your models ?
> 
> yes, I noticed @tianleiwu had a PR (#15005) and I'm checking with him does he plan to get the PR merged? Since he is OOF, do you know why the PR is pending? I will ask our partner from Nvidia to test the model to see whether they encountered the issue that needs the fix.

The PR (#15005) is ready for review. Please help review it. You may need resolve the conflicts later after it is merged. Thanks.

**Comment by chilo-ms** [comment]
> > > Thanks @chilo-ms. @tlwu had a couple of fixes for CUDA Graphs (open PRs) - I was wondering if you needed fixes similar to that while testing your models ?
> > 
> > 
> > yes, I noticed @tianleiwu had a PR (#15005) and I'm checking with him does he plan to get the PR merged? Since he is OOF, do you know why the PR is pending? I will ask our partner from Nvidia to test the model to see whether they encountered the issue that needs the fix.
> 
> The PR (#15005) is ready for review. Please help review it. You may need resolve the conflicts later after it is merged. Thanks.

The PR looks good to me (not reviewing the multi-stream part). Will wait for your PR merged to main first, and then remove the cuda version macro in my PR as well as test it again.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:904]: let's get rid of the CUDA_VERSION >= 10 check

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:1022]: should this be ORT_ENFORCE(isGraphCaptured) or should we return error status instead if IsGraphCaptured() is false?

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:2905]: typo:function

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:1031]: re: lock_guard in compute_func(), will we get rid of that when we start using separate context for each enqueuev2 invocation?  then we need to make a note to make this function thread safe here.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc
Comment (chilo-ms) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:1022]: We did check IsGraphCaptured() before calling ReplayGraph() in [inference_session.cc](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/session/inference_session.cc#L2120).

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc
Comment (chilo-ms) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:1031]: yes, once we have one context per thread implementation, we will get rid of lock_guard in compute_func. Do you want to make this note now in the PR? or do it later once the one context per thread is done?

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc
Comment (jywu-msft) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:1022]: ok i see. i was just commenting on the fact that ReplayGraph() signature returns a Status. so would it be better to return status rather than do an ORT_ENFORCE assertion which throws exception?

Comment (chilo-ms) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:1022]: I think both approaches work. 
Considering CUDA EP also has ORT_ENFORCE [here](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/cuda/cuda_execution_provider.cc#L203), probably make them consistent at this point.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h
Comment (chilo-ms) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h]: removed.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc
Comment (chilo-ms) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: fixed

File: onnxruntime/core/session/inference_session.cc
Comment (hariharans29) [inline onnxruntime/core/session/inference_session.cc:1596]: I guess the comment needs to address the TRT EP as well (i.e.) if ALL graph nodes are assigned to the TRT EP....

File: onnxruntime/core/session/inference_session.cc
Comment (hariharans29) [inline onnxruntime/core/session/inference_session.cc:1641]: Nit: Maybe add a comment for this else to say - currently this code path is for the TRT EP?

File: onnxruntime/core/session/inference_session.cc
Comment (hariharans29) [inline onnxruntime/core/session/inference_session.cc:1601]: What is the expected behavior when a user registers both the TRT and the CUDA EPs (and they set graph capture to true for both EPs) ?
If I understand correctly, the code path below will first look to see if the TRT EP can capture the graph - if it can then all is good, but if it can't then it will return an error status back to the user ? Do you want to check if the CUDA EP can capture the graph before returning the error status to the user ?

Comment (chilo-ms) [reply onnxruntime/core/session/inference_session.cc:1601]: Re:  What is the expected behavior when a user registers both the TRT and the CUDA EPs (and they set graph capture to true for both EPs) ?
That's a good question. My thinking of expected behavior is either (1) all graph nodes assigned to TRT EP and TRT EP will run with cuda graph or (2) only a portion of graph nodes assigned to TRT EP, ORT throws exception to let users know and expect them to disable cuda graph for TRT EP.

Re: Do you want to check if the CUDA EP can capture the graph before returning the error status to the user ?
I think we can still do this, but just wondering whether this will be a real case where users register TRT/CUDA EP with both cuda graph enabled and end up only using CUDA EP with cuda graph. Most users registered TRT EP will want to use TRT EP and also, it's highly possible TRT/CUDA EP without cuda graph has better performance than only CUDA EP with cuda graph feature.

Comment (hariharans29) [reply onnxruntime/core/session/inference_session.cc:1608]: Cool, thanks. If it is not a common scenario, then we probably don't need to address it right now.

**Comment by tianleiwu** [comment]
LGTM. Please resolve conflicts.

**Review by jywu-msft — APPROVED** [review]

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**
